doc/man/mailmail.1
.TH MAILMAIL "1" "July 2003" "" ""
.SH NAME
mailmail \- Twisted sendmail compatibility script
.SH SYNOPSIS
.B mailmail [recipient addresses]
.SH DESCRIPTION
mailmail reads RFC822 message text from standard input and delivers them,
using SMTP, to a Mail Transfer Agent listening at 127.0.0.1:25.  It accepts
(but does not necessarily implement) many of the standard sendmail(1)
options, but it is preferable to list only the recipient addresses on
the command line, and include a \fBFrom\fR header within the message text
indicating the sender.
.SH AUTHOR
Written by Jp Calderone
.SH "REPORTING BUGS"
Report bugs to <twisted-python@twistedmatrix.com>.
.SH COPYRIGHT
Copyright \(co 2003 Matthew W. Lefkowitz
.br
This is free software; see the source for copying conditions.  There is NO
warranty; not even for MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.


NEWS
Ticket numbers in this file can be looked up by visiting
http://twistedmatrix.com/trac/ticket/<number>

Twisted Core 16.0.0pre2 (2016-02-28)
====================================

Features
--------
 - twisted.cred.strports has been ported to Python 3. (#8216)

Deprecations and Removals
-------------------------
 - The __version__ attribute of former subprojects (conch, mail,
   names, news, pair, runner, web, and words) is deprecated in
   preference to the central twisted.__version__. (#8219)

Other
-----
 - #8220


Twisted Conch 16.0.0pre2 (2016-02-28)
=====================================

No significant changes have been made for this release.


Twisted Mail 16.0.0pre2 (2016-02-28)
====================================

No significant changes have been made for this release.


Twisted Names 16.0.0pre2 (2016-02-28)
=====================================

No significant changes have been made for this release.


Twisted News 16.0.0pre2 (2016-02-28)
====================================

No significant changes have been made for this release.


Twisted Pair 16.0.0pre2 (2016-02-28)
====================================

No significant changes have been made for this release.


Twisted Runner 16.0.0pre2 (2016-02-28)
======================================

No significant changes have been made for this release.


Twisted Web 16.0.0pre2 (2016-02-28)
===================================

No significant changes have been made for this release.


Twisted Words 16.0.0pre2 (2016-02-28)
=====================================

No significant changes have been made for this release.


Twisted Core 16.0.0pre1 (2016-02-16)
====================================

Features
--------
 - todo parameter for IReporter.addExpectedSuccess and
   IReporter.addUnexpectedSuccess is no longer required. If not
   provided, a sensible default will be used instead. (#4811)
 - A new string endpoint type, "tls:", allows for properly-verified
   TLS (unlike "ssl:", always matching hostname resolution with
   certificate hostname verification) with faster IPv4/IPv6
   connections.  This comes with an accompanying function,
   twisted.internet.endpoints.wrapClientTLS, which can wrap an
   arbitrary client endpoint with client TLS. (#5642)
 - twisted.python.filepath.makedirs accepts an ignoreExistingDirectory
   flag which ignore the OSError raised by os.makedirs if requested
   directory already exists. (#5704)
 - twisted.protocols.amp has been ported to Python 3. (#6833)
 - twisted.internet.ssl.trustRootFromCertificates returns an object
   suitable for use as trustRoot= to
   twisted.internet.ssl.optionsForClientTLS that trusts multiple
   certificates. (#7671)
 - twisted.python.roots is now ported to Python 3. (#8131)

Bugfixes
--------
 - Expected failures from standard library unittest no longer fail
   with Trial reporters. (#4811)
 - twisted.internet.endpoints.HostnameEndpoint.connect no longer fails
   with an AlreadyCalledError when the Deferred it returns is
   cancelled after all outgoing connection attempts have been made but
   none have yet succeeded or failed. (#8014)
 - twisted.internet.task.LoopingCall.withCount when run with internal
   of 0, now calls the countCallable with 1, regardless of the time
   passed between calls. (#8125)

Improved Documentation
----------------------
 - The Twisted Project has adopted the Contributor Covenant as its
   Code of Conduct. (#8173)

Deprecations and Removals
-------------------------
 - twisted.internet.task.LoopingCall.deferred is now deprecated. Use
   the deferred returned by twisted.internet.task.LoopingCall.start()
   (#8116)
 - twisted.internet.gtkreactor, the GTK+ 1 reactor deprecated since
   Twisted 10.1, has been removed. This does not affect the GTK2,
   GLib, GTK3, or GObject-Introspection reactors. (#8145)
 - twisted.protocols.mice, containing a Logitech MouseMan serial
   driver, has been deprecated. (#8148)

Other
-----
 - #6842, #6978, #7668, #7791, #7881, #7943, #7944, #8050, #8104,
   #8115, #8119, #8122, #8139, #8144, #8154, #8162, #8180, #8187


Twisted Conch 16.0.0pre1 (2016-02-16)
=====================================

Features
--------
 - twisted.conch now uses cryptography instead of PyCrypto for its
   underlying crypto operations. (#7413)
 - twisted.conch.ssh.keys is now ported to Python 3. (#7998)

Bugfixes
--------
 - twisted.conch.ssh.channel.SSHChannel's getPeer and getHost methods
   now return an object which provides IAddress instead of an old-
   style tuple address. (#5999)
 - twisted.conch.endpoint.SSHCommandClientEndpoint, when
   authentication is delegated to an SSH agent, no longer leaves the
   agent connection opened when connection to the server is lost.
   (#8138)

Other
-----
 - #7037, #7715, #8200, #8208


Twisted Mail 16.0.0pre1 (2016-02-16)
====================================

No significant changes have been made for this release.


Twisted Names 16.0.0pre1 (2016-02-16)
=====================================

No significant changes have been made for this release.


Twisted News 16.0.0pre1 (2016-02-16)
====================================

No significant changes have been made for this release.


Twisted Pair 16.0.0pre1 (2016-02-16)
====================================

No significant changes have been made for this release.


Twisted Runner 16.0.0pre1 (2016-02-16)
======================================

No significant changes have been made for this release.


Twisted Web 16.0.0pre1 (2016-02-16)
===================================

Features
--------
 - twisted.web.http_headers._DictHeaders now correctly handles
   updating via keyword arguments in Python 3 (therefore
   twisted.web.http_headers is now fully ported to Python 3). (#6082)
 - twisted.web.wsgi has been ported to Python 3. (#7993)
 - twisted.web.http_headers.Headers now accepts both Unicode and
   bytestring keys and values, encoding to iso-8859-1 and utf8
   respectively. (#8129)
 - twisted.web.vhost ported to Python 3. (#8132)

Bugfixes
--------

twisted/scripts/mailmail.py

# Twisted, the Framework of Your Internet
# Copyright (C) 2001 Matthew W. Lefkowitz
#
# This library is free software; you can redistribute it and/or
# modify it under the terms of version 2.1 of the GNU Lesser General Public
# License as published by the Free Software Foundation.
#
# This library is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
# Lesser General Public License for more details.
#
# You should have received a copy of the GNU Lesser General Public
# License along with this library; if not, write to the Free Software
# Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA
#

"""
Implementation module for the `newtexaco` command.

The name is preliminary and subject to change.
"""

import os
import sys
import rfc822
import socket
import getpass
from ConfigParser import ConfigParser

try:
    import cStringIO as StringIO
except:
    import StringIO

from twisted.internet import reactor
from twisted.protocols import smtp
from twisted.mail import bounce

GLOBAL_CFG = "/etc/mailmail"
LOCAL_CFG = os.path.expanduser("~/.twisted/mailmail")
SMARTHOST = '127.0.0.1'

ERROR_FMT = """\
Subject: Failed Message Delivery

  Message delivery failed.  The following occurred:
  
  %s
-- 
The Twisted sendmail application.
"""

def log(message, *args):
    sys.stderr.write(str(message) % args + '\n')

class Options:
    """
    @type to: C{list} of C{str}
    @ivar to: The addresses to which to deliver this message.
    
    @type sender: C{str}
    @ivar sender: The address from which this message is being sent.
    
    @type body: C{file}
    @ivar body: The object from which the message is to be read.
    """

class JointFile:
    def __init__(self, files, terminateOnPeriod = True):
        self.files = files
        self.files.reverse()
        self.used = []
    
    def read(self, bytes=None):
        r = ''
        while self.files:
            r = self.files[-1].read(bytes)
            if r != '':
                return self.filter(r)
            self.used.append(self.files.pop())
        return ''

    def readline(self):
        if not self.files:
            return ''
        while self.files:
            r = self.files[-1].readline()
            if r != '':
                return self.filter(r)
            self.used.append(self.files.pop())
        return ''
    
    def readlines(self, bytes=None):
        if not self.files:
            return []
        while self.files:
            r = self.files[-1].readlines(bytes)
            if r != []:
                r[-1] == self.filter(r[-1])
                if not r[-1]:
                    del r[-1]
                return r
            self.used.append(self.files.pop())
        return []
    
    def filter(self, s):
        i = s.find('\n.\r')
        if i == -1:
            i = s.find('\n.\n')
        if i == -1:
            return s
        while self.files:
            self.used.append(self.files.pop())
        return s[:i + 1]
    
    def fileno(self):
        return self.files and self.files[-1].fileno() or self.used[-1].fileno()
    
    def close(self):
        for f in self.used:
            f.close()
        for f in self.files:
            f.close()

def getlogin():
    try:
        return os.getlogin()
    except:
        return getpass.getuser()
        

def parseOptions(argv):
    o = Options()
    o.to = [e for e in argv if not e.startswith('-')]
    o.sender = '@'.join((getlogin(), socket.gethostname()))
    
    # Just be very stupid

    # Skip -bm -- it is the default
    
    # -bp lists queue information.  Screw that.
    if '-bp' in argv:
        raise ValueError, "Unsupported option"
    
    # -bs makes sendmail use stdin/stdout as its transport.  Screw that.
    if '-bs' in argv:
        raise ValueError, "Unsupported option"
    
    # -F sets who the mail is from, but is overridable by the From header
    if '-F' in argv:
        o.sender = argv[argv.index('-F') + 1]
        o.to.remove(o.sender)
    
    # -i and -oi makes us ignore lone "."
    if ('-i' in argv) or ('-oi' in argv):
        o.ignoreDot = True
    else:
        o.ignoreDot = False
    
    # -odb is background delivery
    if '-odb' in argv:
        o.background = True
    else:
        o.background = False
    
    # -odf is foreground delivery
    if '-odf' in argv:
        o.background = False
    else:
        o.background = True
    
    # -oem and -em cause errors to be mailed back to the sender.
    # It is also the default.
    
    # -oep and -ep cause errors to be printed to stderr
    if ('-oep' in argv) or ('-ep' in argv):
        o.printErrors = True
    else:
        o.printErrors = False
    
    # -om causes a copy of the message to be sent to the sender if the sender
    # appears in an alias expansion.  We do not support aliases.
    if '-om' in argv:
        raise ValueError, "Unsupported option"

    # -t causes us to pick the recipients of the message from the To, Cc, and Bcc
    # headers, and to remove the Bcc header if present.
    if '-t' in argv:
        o.recipientsFromHeaders = True
        o.excludeAddresses = o.to
        o.to = []
    else:
        o.recipientsFromHeaders = False
        o.exludeAddresses = []
    
    headers = []
    buffer = StringIO.StringIO()
    while 1:
        write = 1
        line = sys.stdin.readline()
        if not line.strip():
            break
        
        hdrs = line.split(': ', 1)
        
        hdr = hdrs[0].lower()
        if o.recipientsFromHeaders and hdr in ('to', 'cc', 'bcc'):
            o.to.extend([
                a[1] for a in rfc822.AddressList(hdrs[1]).addresslist
            ])
            if hdr != 'bcc':
                write = 0
        elif hdr == 'from':
            o.sender = rfc822.parseaddr(hdrs[1])[1]
        
        if write:
            buffer.write(line)
    buffer.write(line)

    if o.recipientsFromHeaders:
        for a in o.excludeAddresses:
            try:
                o.to.remove(a)
            except:
                pass

    buffer.seek(0, 0)
    o.body = JointFile([buffer, sys.stdin], not o.ignoreDot)
    return o

class Configuration:
    """
    @ivar allowUIDs
    @ivar allowGIDs
    @ivar denyUIDs
    @ivar denyGIDs
    @ivar useraccess
    @ivar groupaccess
    @ivar identities
    @ivar smarthost
    @ivar domain
    @ivar defaultAccess
    """
    def __init__(self):
        self.allowUIDs = []
        self.denyUIDs = []
        self.allowGIDs = []
        self.denyGIDs = []
        self.useraccess = 'deny'
        self.groupaccess= 'deny'
        
        self.identities = {}
        self.smarthost = None
        self.domain = None
        
        self.defaultAccess = True

def loadConfig(path):
    # [useraccess]
    # allow=uid1,uid2,...
    # deny=uid1,uid2,...
    # order=allow,deny
    # [groupaccess]
    # allow=gid1,gid2,...
    # deny=gid1,gid2,...
    # order=deny,allow
    # [identity]
    # host1=username:password
    # host2=username:password
    # [addresses]
    # smarthost=a.b.c.d
    # default_domain=x.y.z
    
    c = Configuration()
    
    if not os.access(path, os.R_OK):
        return c

    p = ConfigParser()
    p.read(path)
    
    au = c.allowUIDs
    du = c.denyUIDs
    ag = c.allowGIDs
    dg = c.denyGIDs
    for (section, a, d) in (('useraccess', au, du), ('groupaccess', ag, dg)):
        if p.has_section(section):
            for (mode, L) in (('allow', a), ('deny', d)):
                for id in p.get(section, mode).split(','):
                    try:
                        id = int(id)
                    except ValueError:
                        log("Illegal %sID in [%s] section: %s", section[0].upper(), section, id)
                    else:
                        L.append(id)
            order = p.get(section, 'order')
            order = map(str.split, map(str.lower, order.split(',')))
            if order[0] == 'allow':
                setattr(c, section, 'allow')
            else:
                setattr(c, section, 'deny')

    if p.has_section('identity'):
        for (host, up) in p.items('identity'):
            parts = up.split(':', 1)
            if len(parts) != 2:
                log("Illegal entry in [identity] section: %s", up)
                continue
            p.identities[host] = parts

    if p.has_section('addresses'):
        c.smarthost = p.get('addresses', 'smarthost')
        c.domain = p.get('addresses', 'default_domain')

    return c

def success(result):
    reactor.stop()

failed = None
def failure(f):
    global failed
    reactor.stop()
    failed = f

def sendmail(host, options, ident):
    from twisted.protocols.smtp import sendmail
    d = sendmail(host, options.sender, options.to, options.body)
    d.addCallbacks(success, failure)
    reactor.run()

def senderror(failure, options):
    recipient = [options.sender]
    sender = 'Internally Generated Message (%s)' % (sys.argv[0],)
    error = StringIO.StringIO()
    failure.printTraceback(file=error)
    body = StringIO.StringIO(ERROR_FMT % error.getvalue())

    from twisted.protocols.smtp import sendmail
    d = sendmail('localhost', sender, recipient, body)
    d.addBoth(lambda _: reactor.stop())

def deny(conf):
    uid = os.getuid()
    gid = os.getgid()
    
    if conf.useraccess == 'deny':
        if uid in conf.denyUIDs:
            return True
        if uid in conf.allowUIDs:
            return False
    else:
        if uid in conf.allowUIDs:
            return False
        if uid in conf.denyUIDs:
            return True

    if conf.groupaccess == 'deny':
        if gid in conf.denyGIDs:
            return True
        if gid in conf.allowGIDs:
            return False
    else:
        if gid in conf.allowGIDs:
            return False
        if gid in conf.denyGIDs:
            return True
    
    return not conf.defaultAccess

def run():
    o = parseOptions(sys.argv[1:])
    gConf = loadConfig(GLOBAL_CFG)
    lConf = loadConfig(LOCAL_CFG)
    
    if deny(gConf) or deny(lConf):
        log("Permission denied")
        return

    host = lConf.smarthost or gConf.smarthost or SMARTHOST
    
    ident = gConf.identities.copy()
    ident.update(lConf.identities)
    
    if lConf.domain:
        smtp.DNSNAME = lConf.domain
    elif gConf.domain:
        smtp.DNSNAME = gConf.domain

    sendmail(host, o, ident)

    if failed:
        if o.printErrors:
            failed.printTraceback(file=sys.stderr)
            raise SystemExit(1)
        else:
            senderror(failed, o)
